{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import h5py\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "meta\n",
      "data\n",
      "info\n",
      "auc\n",
      "(11284,)\n",
      "[b'3116748351' b'5069849036' b'4882401623' ... b'5742234142' b'6773612787'\n",
      " b'7613399024']\n",
      "bigwig_file\n",
      "(11284,)\n",
      "[b'3DFF72D2-F292-497E-ACE3-6FAA9C884205.bw'\n",
      " b'B1E54366-42B9-463C-8615-B34D52BD14DC.bw'\n",
      " b'473713F7-EB41-4F20-A37F-ACD209E3CB75.bw' ...\n",
      " b'76987B28-B56B-4A1F-B77C-1E08B8EFEF90.bw'\n",
      " b'33737781-8638-4FA2-AD4C-E888BB9343D8.bw'\n",
      " b'0BB05CA7-C6FF-42A1-919C-D801F471CBBD.bw']\n",
      "cancertype\n",
      "(11284,)\n",
      "[b'Liver Hepatocellular Carcinoma' b'Prostate Adenocarcinoma'\n",
      " b'Rectum Adenocarcinoma' ... b'Brain Lower Grade Glioma'\n",
      " b'Acute Myeloid Leukemia' b'Thyroid Carcinoma']\n",
      "gdc_annotations\n",
      "(11284,)\n",
      "[b'NULL' b'NULL' b'NULL' ... b'NULL'\n",
      " b'list(category = \"Alternate sample pipeline\", status = \"Approved\", entity_id = \"fff35c80-88cd-4923-80c1-0273ba5bed0f\", classification = \"Notification\", updated_datetime = \"2016-05-01T15:00:21.638875-05:00\", created_datetime = \"2012-11-13T12:37:22-05:00\", entity_submitter_id = \"TCGA-AB-2881\", notes = \"Biospecimens from this case belong to Batch 25 and were processed into analyte outside of the normal TCGA standardized laboratory pipeline.  Users should note that the resulting characterization data may exhibit features (e.g. batch effects) due to this difference.\", \\n    entity_type = \"case\", submitter_id = \"12189\", case_id = \"fff35c80-88cd-4923-80c1-0273ba5bed0f\", annotation_id = \"91e80142-dccb-546b-838d-03837dd1a30e\")'\n",
      " b'NULL']\n",
      "gdc_cases.annotations\n",
      "(11284,)\n",
      "[b'NULL' b'NULL' b'NULL' ... b'NULL'\n",
      " b'list(category = \"Alternate sample pipeline\", status = \"Approved\", classification = \"Notification\", updated_datetime = \"2016-05-01T15:00:21.638875-05:00\", created_datetime = \"2012-11-13T12:37:22-05:00\", notes = \"Biospecimens from this case belong to Batch 25 and were processed into analyte outside of the normal TCGA standardized laboratory pipeline.  Users should note that the resulting characterization data may exhibit features (e.g. batch effects) due to this difference.\", submitter_id = \"12189\", \\n    annotation_id = \"91e80142-dccb-546b-838d-03837dd1a30e\")'\n",
      " b'NULL']\n",
      "gdc_cases.case_id\n",
      "(11284,)\n",
      "[b'0004d251-3f70-4395-b175-c94c2f5b1b81'\n",
      " b'000d566c-96c7-4f1c-b36e-fa2222467983'\n",
      " b'0011a67b-1ba9-4a32-a6b8-7850759a38cf' ...\n",
      " b'fff304a2-113f-499d-a88c-9d3660c348d9'\n",
      " b'fff35c80-88cd-4923-80c1-0273ba5bed0f'\n",
      " b'fffdb1d9-58d1-425c-ac12-1e1e5f443bf7']\n",
      "gdc_cases.project.name\n",
      "(11284,)\n",
      "[b'Liver Hepatocellular Carcinoma' b'Prostate Adenocarcinoma'\n",
      " b'Rectum Adenocarcinoma' ... b'Brain Lower Grade Glioma'\n",
      " b'Acute Myeloid Leukemia' b'Thyroid Carcinoma']\n",
      "gdc_cases.project.primary_site\n",
      "(11284,)\n",
      "[b'Liver' b'Prostate' b'Colorectal' ... b'Brain' b'Bone Marrow' b'Thyroid']\n",
      "gdc_cases.project.project_id\n",
      "(11284,)\n",
      "[b'TCGA-LIHC' b'TCGA-PRAD' b'TCGA-READ' ... b'TCGA-LGG' b'TCGA-LAML'\n",
      " b'TCGA-THCA']\n",
      "gdc_cases.samples.is_ffpe\n",
      "(11284,)\n",
      "[b'FALSE' b'FALSE' b'FALSE' ... b'FALSE' b'FALSE' b'FALSE']\n",
      "gdc_cases.samples.portions.analytes.aliquots.aliquot_id\n",
      "(11284,)\n",
      "[b'6c3a4f09-1862-4fed-9977-759ec00a5aeb'\n",
      " b'9d5f0b9f-ff36-4a80-8472-412d0ec803c5'\n",
      " b'fbf67b9e-05a7-4280-a7b8-13363f1bd1c4' ...\n",
      " b'1d7d22de-0bfc-4357-8c5a-7289aa0d97b8'\n",
      " b'af389bfe-c693-4763-a719-8f74f265e6ae'\n",
      " b'e9eda4da-4b9d-4add-8229-985b27f5ac2c']\n",
      "gdc_cases.samples.portions.analytes.aliquots.annotations\n",
      "(11284,)\n",
      "[b'NULL' b'NULL' b'NULL' ... b'NULL' b'NULL' b'NULL']\n",
      "gdc_cases.samples.portions.analytes.aliquots.source_center\n",
      "(11284,)\n",
      "[b'23' b'23' b'22' ... b'23' b'23' b'23']\n",
      "gdc_cases.samples.portions.analytes.aliquots.submitter_id\n",
      "(11284,)\n",
      "[b'TCGA-DD-AAVP-01A-11R-A41C-07' b'TCGA-KK-A7B2-01A-12R-A32O-07'\n",
      " b'TCGA-DC-6158-01A-11R-1660-07' ... b'TCGA-RY-A847-01A-11R-A36H-07'\n",
      " b'TCGA-AB-2881-03A-01T-0735-13' b'TCGA-DJ-A4V0-01A-11R-A250-07']\n",
      "gdc_cases.samples.portions.analytes.analyte_id\n",
      "(11284,)\n",
      "[b'2246af47-8d57-49bc-8249-85b5dce8f42c'\n",
      " b'9e093595-180c-4a5c-940e-27ddea8374b3'\n",
      " b'867d8ba8-2261-4c57-9bc9-cb81f7226991' ...\n",
      " b'e2c8613e-f156-4ab7-af82-321dbdb9dc41'\n",
      " b'9cced397-aa3f-4451-a942-54528d5e8b1e'\n",
      " b'1624438d-5d9f-499d-84a1-14394d5228a1']\n",
      "gdc_cases.samples.portions.analytes.analyte_type\n",
      "(11284,)\n",
      "[b'RNA' b'RNA' b'RNA' ... b'RNA' b'Total RNA' b'RNA']\n",
      "gdc_cases.samples.portions.analytes.analyte_type_id\n",
      "(11284,)\n",
      "[b'R' b'R' b'R' ... b'R' b'T' b'R']\n",
      "gdc_cases.samples.portions.analytes.annotations\n",
      "(11284,)\n",
      "[b'NULL' b'NULL' b'NULL' ... b'NULL' b'NULL' b'NULL']\n",
      "gdc_cases.samples.portions.analytes.submitter_id\n",
      "(11284,)\n",
      "[b'TCGA-DD-AAVP-01A-11R' b'TCGA-KK-A7B2-01A-12R' b'TCGA-DC-6158-01A-11R'\n",
      " ... b'TCGA-RY-A847-01A-11R' b'TCGA-AB-2881-03A-01T'\n",
      " b'TCGA-DJ-A4V0-01A-11R']\n",
      "gdc_cases.samples.portions.creation_datetime\n",
      "(11284,)\n",
      "[b'1397088000' b'1377475200' b'1299196800' ... b'1382918400' b'0'\n",
      " b'1351641600']\n",
      "gdc_cases.samples.portions.portion_id\n",
      "(11284,)\n",
      "[b'671ac90c-63b4-450b-814b-797349f81f15'\n",
      " b'ab77d27b-5d3d-46d7-9b19-30aef7fc34d0'\n",
      " b'abc0ddaf-1842-4e7d-bca3-c54a91d54170' ...\n",
      " b'7a9f863a-d7fe-476a-93ff-3c9b95e597f7'\n",
      " b'71be866a-c0a1-447e-94de-ef2a85a891c7'\n",
      " b'e78cb405-c721-4a2c-a099-7c8a4eb5e46d']\n",
      "gdc_cases.samples.portions.portion_number\n",
      "(11284,)\n",
      "[b'11' b'12' b'11' ... b'11' b'01' b'11']\n",
      "gdc_cases.samples.portions.slides\n",
      "(11284,)\n",
      "[b'list(percent_tumor_nuclei = 80, percent_monocyte_infiltration = 1, percent_normal_cells = 0, percent_stromal_cells = 20, updated_datetime = \"2016-09-08T11:04:30.952291-05:00\", created_datetime = NA, percent_eosinophil_infiltration = NA, percent_lymphocyte_infiltration = 2, percent_neutrophil_infiltration = 1, state = NA, section_location = \"TOP\", percent_granulocyte_infiltration = NA, percent_necrosis = 0, slide_id = \"c1eab3d0-509c-411b-a23d-19b2515c781d\", percent_inflam_infiltration = NA, submitter_id = \"TCGA-DD-AAVP-01A-01-TS1\", \\n    number_proliferating_cells = NA, percent_tumor_cells = 80)'\n",
      " b'list(percent_tumor_nuclei = 80, percent_monocyte_infiltration = 2, percent_normal_cells = 0, percent_stromal_cells = 55, updated_datetime = \"2016-09-08T13:59:46.086986-05:00\", created_datetime = NA, percent_eosinophil_infiltration = NA, percent_lymphocyte_infiltration = 10, percent_neutrophil_infiltration = 2, state = NA, section_location = \"TOP\", percent_granulocyte_infiltration = NA, percent_necrosis = 1, slide_id = \"3f1e3102-294c-47f4-a1da-1d0f09abe9c1\", percent_inflam_infiltration = NA, submitter_id = \"TCGA-KK-A7B2-01A-01-TSA\", \\n    number_proliferating_cells = NA, percent_tumor_cells = 44)'\n",
      " b'list(percent_tumor_nuclei = c(70, 70), percent_monocyte_infiltration = c(NA, NA), percent_normal_cells = c(10, 8), percent_stromal_cells = c(20, 20), updated_datetime = c(\"2016-09-08T10:40:20.418685-05:00\", \"2016-09-08T10:40:20.418685-05:00\"), created_datetime = c(NA, NA), percent_eosinophil_infiltration = c(NA, NA), percent_lymphocyte_infiltration = c(NA, NA), percent_neutrophil_infiltration = c(NA, NA), state = c(NA, NA), section_location = c(\"TOP\", \"BOTTOM\"), percent_granulocyte_infiltration = c(NA, \\nNA), percent_necrosis = c(5, 5), slide_id = c(\"f5f1dac8-1047-478e-bc77-5c7ddd43566f\", \"b4e1dde1-d2b8-4dd2-b0c0-5bb927825869\"), percent_inflam_infiltration = c(NA, NA), submitter_id = c(\"TCGA-DC-6158-01A-01-TS1\", \"TCGA-DC-6158-01A-01-BS1\"), number_proliferating_cells = c(NA, NA), percent_tumor_cells = c(65, 67))'\n",
      " ...\n",
      " b'list(percent_tumor_nuclei = 75, percent_monocyte_infiltration = 0, percent_normal_cells = 10, percent_stromal_cells = 65, updated_datetime = \"2016-09-08T12:43:40.136159-05:00\", created_datetime = NA, percent_eosinophil_infiltration = NA, percent_lymphocyte_infiltration = 0, percent_neutrophil_infiltration = 0, state = NA, section_location = \"TOP\", percent_granulocyte_infiltration = NA, percent_necrosis = 0, slide_id = \"0d4e8888-4d3a-402f-8db8-495e9e366832\", percent_inflam_infiltration = NA, submitter_id = \"TCGA-RY-A847-01A-01-TS1\", \\n    number_proliferating_cells = NA, percent_tumor_cells = 25)'\n",
      " b'NULL'\n",
      " b'list(percent_tumor_nuclei = 75, percent_monocyte_infiltration = 1, percent_normal_cells = 2, percent_stromal_cells = 23, updated_datetime = \"2016-09-08T13:40:21.455089-05:00\", created_datetime = NA, percent_eosinophil_infiltration = NA, percent_lymphocyte_infiltration = 2, percent_neutrophil_infiltration = 0, state = NA, section_location = \"TOP\", percent_granulocyte_infiltration = NA, percent_necrosis = 0, slide_id = \"3fd33092-6128-4444-b9d3-0ac11627eff9\", percent_inflam_infiltration = NA, submitter_id = \"TCGA-DJ-A4V0-01A-01-TSA\", \\n    number_proliferating_cells = NA, percent_tumor_cells = 75)']\n",
      "gdc_cases.samples.portions.submitter_id\n",
      "(11284,)\n",
      "[b'TCGA-DD-AAVP-01A-11' b'TCGA-KK-A7B2-01A-12' b'TCGA-DC-6158-01A-11' ...\n",
      " b'TCGA-RY-A847-01A-11' b'TCGA-AB-2881-03A-01' b'TCGA-DJ-A4V0-01A-11']\n",
      "gdc_cases.samples.sample_id\n",
      "(11284,)\n",
      "[b'9259e9ee-7279-4b62-8512-509cb705029c'\n",
      " b'bccb7f9d-8dde-45b7-a50d-6c7233239ceb'\n",
      " b'03877fe0-b0d6-4800-be46-b45c70350e48' ...\n",
      " b'bf8e8426-dc44-40b0-8459-ad1dbaf9ba5a'\n",
      " b'421c391a-33c7-4a6d-aecb-9a1cb4e8a2d9'\n",
      " b'a418a2e8-6a8d-4e7d-8c43-f09b521e51d8']\n",
      "gdc_cases.samples.sample_type\n",
      "(11284,)\n",
      "[b'Primary Tumor' b'Primary Tumor' b'Primary Tumor' ... b'Primary Tumor'\n",
      " b'Primary Blood Derived Cancer - Peripheral Blood' b'Primary Tumor']\n",
      "gdc_cases.samples.sample_type_id\n",
      "(11284,)\n",
      "[b'01' b'01' b'01' ... b'01' b'03' b'01']\n",
      "gdc_cases.samples.submitter_id\n",
      "(11284,)\n",
      "[b'TCGA-DD-AAVP-01A' b'TCGA-KK-A7B2-01A' b'TCGA-DC-6158-01A' ...\n",
      " b'TCGA-RY-A847-01A' b'TCGA-AB-2881-03A' b'TCGA-DJ-A4V0-01A']\n",
      "gdc_cases.submitter_id\n",
      "(11284,)\n",
      "[b'TCGA-DD-AAVP' b'TCGA-KK-A7B2' b'TCGA-DC-6158' ... b'TCGA-RY-A847'\n",
      " b'TCGA-AB-2881' b'TCGA-DJ-A4V0']\n",
      "gdc_cases.tissue_source_site.bcr_id\n",
      "(11284,)\n",
      "[b'NCH' b'IGC' b'IGC' ... b'NCH' b'NCH' b'NCH']\n",
      "gdc_cases.tissue_source_site.name\n",
      "(11284,)\n",
      "[b'Mayo Clinic - Rochester' b'MD Anderson Cancer Center' b'MSKCC' ...\n",
      " b'University of California San Francisco' b'Washington University'\n",
      " b'Memorial Sloan Kettering']\n",
      "gdc_cases.tissue_source_site.project\n",
      "(11284,)\n",
      "[b'Liver hepatocellular carcinoma' b'Prostate adenocarcinoma'\n",
      " b'Rectum adenocarcinoma' ... b'Brain Lower Grade Glioma'\n",
      " b'Acute Myeloid Leukemia' b'Thyroid carcinoma']\n",
      "gdc_cases.tissue_source_site.tissue_source_site_id\n",
      "(11284,)\n",
      "[b'c2eb5c30-93cc-5a99-b9e4-f676e35ed7f8'\n",
      " b'c4a03d2d-b7e9-5506-9521-4df06091fb20'\n",
      " b'2ec37cf9-8bca-524f-9fb8-1241ab73fac7' ...\n",
      " b'02c87f8b-69ba-51fb-a99f-c2f22183b992'\n",
      " b'5b2bb0b1-686a-59fa-9f1c-960a8d701e83'\n",
      " b'612b370d-47ee-5e52-ab25-6770f395998f']\n",
      "gdc_cases.updated_datetime\n",
      "(11284,)\n",
      "[b'2016-09-08T11:04:30.952291-05:00' b'2016-09-08T13:59:46.086986-05:00'\n",
      " b'2016-09-08T10:40:20.418685-05:00' ...\n",
      " b'2016-09-08T12:43:40.136159-05:00' b'2016-09-08T13:13:12.360689-05:00'\n",
      " b'2016-09-08T13:40:21.455089-05:00']\n",
      "gdc_center.center_id\n",
      "(11284,)\n",
      "[b'ee7a85b3-8177-5d60-a10c-51180eb9009c'\n",
      " b'ee7a85b3-8177-5d60-a10c-51180eb9009c'\n",
      " b'ee7a85b3-8177-5d60-a10c-51180eb9009c' ...\n",
      " b'ee7a85b3-8177-5d60-a10c-51180eb9009c'\n",
      " b'6eba705a-0f00-5aa2-b1d0-04dbf62100cc'\n",
      " b'ee7a85b3-8177-5d60-a10c-51180eb9009c']\n",
      "gdc_center.code\n",
      "(11284,)\n",
      "[b'07' b'07' b'07' ... b'07' b'13' b'07']\n",
      "gdc_center.name\n",
      "(11284,)\n",
      "[b'University of North Carolina' b'University of North Carolina'\n",
      " b'University of North Carolina' ... b'University of North Carolina'\n",
      " b\"Canada's Michael Smith Genome Sciences Centre\"\n",
      " b'University of North Carolina']\n",
      "gdc_center.namespace\n",
      "(11284,)\n",
      "[b'unc.edu' b'unc.edu' b'unc.edu' ... b'unc.edu' b'bcgsc.ca' b'unc.edu']\n",
      "gdc_center.short_name\n",
      "(11284,)\n",
      "[b'UNC' b'UNC' b'UNC' ... b'UNC' b'BCGSC' b'UNC']\n",
      "gdc_data_format\n",
      "(11284,)\n",
      "[b'TARGZ' b'TARGZ' b'TARGZ' ... b'TARGZ' b'TAR' b'TARGZ']\n",
      "gdc_file_id\n",
      "(11284,)\n",
      "[b'3dff72d2-f292-497e-ace3-6faa9c884205'\n",
      " b'b1e54366-42b9-463c-8615-b34d52bd14dc'\n",
      " b'473713f7-eb41-4f20-a37f-acd209e3cb75' ...\n",
      " b'76987b28-b56b-4a1f-b77c-1e08b8efef90'\n",
      " b'33737781-8638-4fa2-ad4c-e888bb9343d8'\n",
      " b'0bb05ca7-c6ff-42a1-919c-d801f471cbbd']\n",
      "gdc_file_name\n",
      "(11284,)\n",
      "[b'UNCID_2660992.6c3a4f09-1862-4fed-9977-759ec00a5aeb.140826_UNC9-SN296_0433_BC5FWDACXX_5_GGCTAC.tar.gz'\n",
      " b'UNCID_2322574.9d5f0b9f-ff36-4a80-8472-412d0ec803c5.131021_UNC9-SN296_0409_BC2M6NACXX_4_ATCACG.tar.gz'\n",
      " b'UNCID_2185604.fbf67b9e-05a7-4280-a7b8-13363f1bd1c4.121211_UNC16-SN851_0199_BD1KLUACXX_3_AGTTCC.tar.gz'\n",
      " ...\n",
      " b'UNCID_2520250.1d7d22de-0bfc-4357-8c5a-7289aa0d97b8.140325_UNC15-SN850_0358_AC3LP7ACXX_3_ACAGTG.tar.gz'\n",
      " b'TCGA-AB-2881-03A-01T-0735-13_rnaseq_fastq.tar'\n",
      " b'UNCID_2184304.e9eda4da-4b9d-4add-8229-985b27f5ac2c.130214_UNC11-SN627_0281_AD1UHCACXX_7_ACTTGA.tar.gz']\n",
      "gdc_file_size\n",
      "(11284,)\n",
      "[b'2726695328' b'5351542755' b'4518948272' ... b'5653270016' b'5393807360'\n",
      " b'6577268060']\n",
      "gdc_md5sum\n",
      "(11284,)\n",
      "[b'9870f011553dcbea40fa35a43e4beac4' b'903fa5aa4470dfdf2b1b49fb63f33cc7'\n",
      " b'8bb79b9d0b7387fd617ac9a5cf9dc0f4' ...\n",
      " b'97e33728f3aa9e2f649f2cfdf58210eb' b'3fcf2c40f11d5143fb980012e5843d6d'\n",
      " b'c4073b2e1452bb5f8d7a731ef278e5f4']\n",
      "gdc_metadata_files.created_datetime.analysis\n",
      "(11284,)\n",
      "[b'2016-06-22T11:15:44.245302-05:00' b'2016-05-03T11:30:02.988971-05:00'\n",
      " b'2016-05-03T10:59:42.133373-05:00' ...\n",
      " b'2016-05-03T11:51:27.517228-05:00' b'2016-05-03T11:38:01.265665-05:00'\n",
      " b'2016-05-03T10:58:56.650557-05:00']\n",
      "gdc_metadata_files.file_id.analysis\n",
      "(11284,)\n",
      "[b'ebc35995-75c4-4dd6-b895-7153d04f27f9'\n",
      " b'9ae5c78f-8484-4edd-93a8-214a6d4de54e'\n",
      " b'29433d82-efbc-4d1f-9719-10a398da017d' ...\n",
      " b'473fa97a-9ea5-4758-82ee-524a0bceaca7'\n",
      " b'8d4bdcc3-232f-4b80-94f4-dd0bf1e44d65'\n",
      " b'd0e14c59-fa4c-47ba-96a9-80d6b5e18f10']\n",
      "gdc_metadata_files.file_id.experiment\n",
      "(11284,)\n",
      "[b'dbc3f41a-b742-4fb8-8648-69a99a215f78'\n",
      " b'2da96e5c-fb40-458c-8a16-b69248a67872'\n",
      " b'a8a92d7e-c2b5-418f-aa9c-490327c1ab77' ...\n",
      " b'1d3cbeee-5c1f-4daa-a7d6-5220ee370f4f'\n",
      " b'50d61ac3-6161-48be-9b97-888b4f027a87'\n",
      " b'2007eebf-67b9-411c-ac29-e0779a549e5e']\n",
      "gdc_metadata_files.file_id.run\n",
      "(11284,)\n",
      "[b'01bd9600-3510-49c7-a474-5e7e5fcc3414'\n",
      " b'47a46d53-de20-497d-b0d9-4b82f70b13a5'\n",
      " b'a5c90d99-90af-4c39-8b7a-d4a2b1613b1c' ...\n",
      " b'e80d0197-46f8-483e-9f18-df28ab8e823a'\n",
      " b'6b8b7430-898f-4cce-8189-19eb1094a2e6'\n",
      " b'764508a2-3d64-4a98-8d21-70de9e3309a4']\n",
      "gdc_metadata_files.file_name.analysis\n",
      "(11284,)\n",
      "[b'62dca2cf-5330-4d57-b322-ed57dc484550_analysis.xml'\n",
      " b'dc2e7732-07a8-42de-86b4-3dc5503c216e_analysis.xml'\n",
      " b'6c36a436-2ed6-49a0-a498-88a48503136e_analysis.xml' ...\n",
      " b'89ac6022-a529-4b81-93d9-fa851c8ee435_analysis.xml'\n",
      " b'44b39039-2cce-402c-bc94-671c306a8f08_analysis.xml'\n",
      " b'2c200aa9-7df8-4469-a421-db4334b7f338_analysis.xml']\n",
      "gdc_metadata_files.file_name.experiment\n",
      "(11284,)\n",
      "[b'62dca2cf-5330-4d57-b322-ed57dc484550_experiment.xml'\n",
      " b'dc2e7732-07a8-42de-86b4-3dc5503c216e_experiment.xml'\n",
      " b'6c36a436-2ed6-49a0-a498-88a48503136e_experiment.xml' ...\n",
      " b'89ac6022-a529-4b81-93d9-fa851c8ee435_experiment.xml'\n",
      " b'44b39039-2cce-402c-bc94-671c306a8f08_experiment.xml'\n",
      " b'2c200aa9-7df8-4469-a421-db4334b7f338_experiment.xml']\n",
      "gdc_metadata_files.file_name.run\n",
      "(11284,)\n",
      "[b'62dca2cf-5330-4d57-b322-ed57dc484550_run.xml'\n",
      " b'dc2e7732-07a8-42de-86b4-3dc5503c216e_run.xml'\n",
      " b'6c36a436-2ed6-49a0-a498-88a48503136e_run.xml' ...\n",
      " b'89ac6022-a529-4b81-93d9-fa851c8ee435_run.xml'\n",
      " b'44b39039-2cce-402c-bc94-671c306a8f08_run.xml'\n",
      " b'2c200aa9-7df8-4469-a421-db4334b7f338_run.xml']\n",
      "gdc_metadata_files.file_size.analysis\n",
      "(11284,)\n",
      "[b'2122' b'2122' b'2123' ... b'2124' b'2284' b'2123']\n",
      "gdc_metadata_files.file_size.experiment\n",
      "(11284,)\n",
      "[b'2628' b'2628' b'2431' ... b'2628' b'3938' b'2431']\n",
      "gdc_metadata_files.file_size.run\n",
      "(11284,)\n",
      "[b'354' b'354' b'354' ... b'354' b'3407' b'354']\n",
      "gdc_metadata_files.md5sum.analysis\n",
      "(11284,)\n",
      "[b'a71e168607d7a29685f654364a7631c5' b'f6549d14d438410af3212e514279a449'\n",
      " b'4b8e2d6dbdb5d83553b907995f482f24' ...\n",
      " b'26fd5c4a973e04c84d4a6b2ed8336b42' b'1ac10804b15fdb370987faa8446b9ee8'\n",
      " b'71b30d03d804238e8db28ceec332960b']\n",
      "gdc_metadata_files.md5sum.experiment\n",
      "(11284,)\n",
      "[b'7149a679ae2898df623b65d7807030dd' b'4e37fd2949af26989d268b44df3e4cdb'\n",
      " b'1bc2281ecb773b96f66f52c1279c10b3' ...\n",
      " b'6c83ad67e1bba307cc248624acd4d62b' b'88ef523ff48c6a35957b9ea8fdb098ad'\n",
      " b'b0d8a45d2090db49b0806056c915bf7f']\n",
      "gdc_metadata_files.md5sum.run\n",
      "(11284,)\n",
      "[b'ca0ec5262de5a6f211febe2f3ad74eca' b'fe4b20e0939b2640912fe846ee5e6b33'\n",
      " b'4c82b64deba5f35b8e29da5b90eb79ee' ...\n",
      " b'a6bfbc9862b3926cddafdc807c03b6ec' b'c4b519cfd319ab127627a608182565f8'\n",
      " b'0fe7555b28f656b22292e378a23c686e']\n",
      "gdc_metadata_files.updated_datetime.analysis\n",
      "(11284,)\n",
      "[b'2016-09-29T23:50:31.404039-05:00' b'2016-09-29T21:40:10.807547-05:00'\n",
      " b'2016-09-29T18:37:16.469086-05:00' ...\n",
      " b'2016-09-29T19:24:43.821317-05:00' b'2016-09-29T21:18:01.322684-05:00'\n",
      " b'2016-09-29T23:07:14.783439-05:00']\n",
      "gdc_metadata_files.updated_datetime.experiment\n",
      "(11284,)\n",
      "[b'2016-08-13T06:57:44.873718-05:00' b'2016-10-07T16:27:24.893743-05:00'\n",
      " b'2016-09-29T22:02:06.782908-05:00' ...\n",
      " b'2016-08-28T02:33:15.912268-05:00' b'2016-08-27T19:46:42.557267-05:00'\n",
      " b'2016-09-29T18:22:41.660257-05:00']\n",
      "gdc_metadata_files.updated_datetime.run\n",
      "(11284,)\n",
      "[b'2016-09-29T17:34:29.260368-05:00' b'2016-09-29T19:25:21.903235-05:00'\n",
      " b'2016-09-29T21:57:26.310154-05:00' ...\n",
      " b'2016-09-29T23:44:24.815358-05:00' b'2016-09-29T20:23:41.491977-05:00'\n",
      " b'2016-09-29T20:41:14.561790-05:00']\n",
      "gdc_platform\n",
      "(11284,)\n",
      "[b'Illumina HiSeq' b'Illumina HiSeq' b'Illumina HiSeq' ...\n",
      " b'Illumina HiSeq' b'Illumina GA' b'Illumina HiSeq']\n",
      "gdc_submitter_id\n",
      "(11284,)\n",
      "[b'62dca2cf-5330-4d57-b322-ed57dc484550'\n",
      " b'dc2e7732-07a8-42de-86b4-3dc5503c216e'\n",
      " b'6c36a436-2ed6-49a0-a498-88a48503136e' ...\n",
      " b'89ac6022-a529-4b81-93d9-fa851c8ee435'\n",
      " b'44b39039-2cce-402c-bc94-671c306a8f08'\n",
      " b'2c200aa9-7df8-4469-a421-db4334b7f338']\n",
      "gdc_updated_datetime\n",
      "(11284,)\n",
      "[b'2016-08-09T14:42:05.417980-05:00' b'2016-08-26T04:54:14.470936-05:00'\n",
      " b'2016-08-09T13:09:58.664649-05:00' ...\n",
      " b'2016-08-25T17:41:13.293379-05:00' b'2016-08-25T14:00:52.797929-05:00'\n",
      " b'2016-08-05T01:07:11.185164-05:00']\n",
      "genes\n",
      "(25150,)\n",
      "[b'A1BG' b'A1BG-AS1' b'A1CF' ... b'ZYX' b'ZZEF1' b'ZZZ3']\n",
      "mapped_read_count\n",
      "(11284,)\n",
      "[b'65175027' b'106227036' b'102592885' ... b'120236713' b'137419287'\n",
      " b'159646577']\n",
      "reads_downloaded\n",
      "(11284,)\n",
      "[b'65993288' b'108023394' b'103976586' ... b'122325822' b'155307816'\n",
      " b'161782584']\n",
      "sampleid\n",
      "(11284,)\n",
      "[b'3DFF72D2-F292-497E-ACE3-6FAA9C884205'\n",
      " b'B1E54366-42B9-463C-8615-B34D52BD14DC'\n",
      " b'473713F7-EB41-4F20-A37F-ACD209E3CB75' ...\n",
      " b'76987B28-B56B-4A1F-B77C-1E08B8EFEF90'\n",
      " b'33737781-8638-4FA2-AD4C-E888BB9343D8'\n",
      " b'0BB05CA7-C6FF-42A1-919C-D801F471CBBD']\n",
      "expression\n",
      "(11284, 25150)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1454762  439935  316636 ...  302428   29863   44717]\n",
      " [  28886   23917     192 ... 1009426  156109  108371]\n",
      " [   4322    9719  154254 ... 1319876  348923  253112]\n",
      " ...\n",
      " [  10570   16608     220 ...  402715  307415   85944]\n",
      " [  28190   52035     157 ... 1888391  660036  369616]\n",
      " [  40426   40053      48 ...  793314  481256  313210]]\n",
      "author\n",
      "(1,)\n",
      "[b'Alexander Lachmann']\n",
      "contact\n",
      "(1,)\n",
      "[b'alexander.lachmann@mssm.edu']\n",
      "countsource\n",
      "(1,)\n",
      "[b'recount2: https://jhubiostatistics.shinyapps.io/recount/']\n",
      "creationdate\n",
      "(1,)\n",
      "[b'10/9/2017']\n",
      "lab\n",
      "(1,)\n",
      "[b\"Ma'ayan Lab, Icahn School of Medicine\"]\n"
     ]
    }
   ],
   "source": [
    "#Importing TCGA RNA expression data downloaded from  https://amp.pharm.mssm.edu/archs4/download.html\n",
    "filename = \"tcga_matrix.h5\"\n",
    "\n",
    "h5 = h5py.File(filename,'r')\n",
    "for k in h5.keys():\n",
    "    print(k)\n",
    "\n",
    "\n",
    "meta = h5['meta']  \n",
    "dat = h5['data']  \n",
    "info = h5['info']  \n",
    "\n",
    "\n",
    "\n",
    "#Showing which type of data it contains    \n",
    "for s in meta.keys():\n",
    "    print(s)\n",
    "    print(meta[s].shape)\n",
    "    print(meta[s].value)\n",
    "    \n",
    "for s in dat.keys():\n",
    "    print(s)\n",
    "    print(dat[s].shape)\n",
    "    print(dat[s].value)    \n",
    "    \n",
    "for s in info.keys():\n",
    "    print(s)\n",
    "    print(info[s].shape)\n",
    "    print(info[s].value)    \n",
    "\n",
    "#h5.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>192</th>\n",
       "      <th>193</th>\n",
       "      <th>194</th>\n",
       "      <th>195</th>\n",
       "      <th>196</th>\n",
       "      <th>197</th>\n",
       "      <th>198</th>\n",
       "      <th>199</th>\n",
       "      <th>200</th>\n",
       "      <th>201</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "      <td>http://www.broadinstitute.org/gsea/msigdb/card...</td>\n",
       "      <td>JUNB</td>\n",
       "      <td>CXCL2</td>\n",
       "      <td>ATF3</td>\n",
       "      <td>NFKBIA</td>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>PTGS2</td>\n",
       "      <td>CXCL1</td>\n",
       "      <td>IER3</td>\n",
       "      <td>...</td>\n",
       "      <td>EIF1</td>\n",
       "      <td>BMP2</td>\n",
       "      <td>DUSP4</td>\n",
       "      <td>PDLIM5</td>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>GFPT2</td>\n",
       "      <td>KLF2</td>\n",
       "      <td>TNC</td>\n",
       "      <td>SERPINB8</td>\n",
       "      <td>MXD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HALLMARK_HYPOXIA</td>\n",
       "      <td>http://www.broadinstitute.org/gsea/msigdb/card...</td>\n",
       "      <td>PGK1</td>\n",
       "      <td>PDK1</td>\n",
       "      <td>GBE1</td>\n",
       "      <td>PFKL</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>ENO2</td>\n",
       "      <td>PGM1</td>\n",
       "      <td>NDRG1</td>\n",
       "      <td>...</td>\n",
       "      <td>HDLBP</td>\n",
       "      <td>ILVBL</td>\n",
       "      <td>NCAN</td>\n",
       "      <td>TGM2</td>\n",
       "      <td>ETS1</td>\n",
       "      <td>HOXB9</td>\n",
       "      <td>SELENBP1</td>\n",
       "      <td>FOSL2</td>\n",
       "      <td>SULT2B1</td>\n",
       "      <td>TGFB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HALLMARK_CHOLESTEROL_HOMEOSTASIS</td>\n",
       "      <td>http://www.broadinstitute.org/gsea/msigdb/card...</td>\n",
       "      <td>FDPS</td>\n",
       "      <td>CYP51A1</td>\n",
       "      <td>IDI1</td>\n",
       "      <td>FDFT1</td>\n",
       "      <td>DHCR7</td>\n",
       "      <td>SQLE</td>\n",
       "      <td>HMGCS1</td>\n",
       "      <td>NSDHL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HALLMARK_MITOTIC_SPINDLE</td>\n",
       "      <td>http://www.broadinstitute.org/gsea/msigdb/card...</td>\n",
       "      <td>ARHGEF2</td>\n",
       "      <td>CLASP1</td>\n",
       "      <td>KIF11</td>\n",
       "      <td>KIF23</td>\n",
       "      <td>ALS2</td>\n",
       "      <td>ARF6</td>\n",
       "      <td>MYO9B</td>\n",
       "      <td>MYH9</td>\n",
       "      <td>...</td>\n",
       "      <td>RHOT2</td>\n",
       "      <td>SORBS2</td>\n",
       "      <td>CDC42EP1</td>\n",
       "      <td>VCL</td>\n",
       "      <td>CLIP1</td>\n",
       "      <td>STK38L</td>\n",
       "      <td>YWHAE</td>\n",
       "      <td>RAPGEF5</td>\n",
       "      <td>CEP72</td>\n",
       "      <td>CSNK1D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HALLMARK_WNT_BETA_CATENIN_SIGNALING</td>\n",
       "      <td>http://www.broadinstitute.org/gsea/msigdb/card...</td>\n",
       "      <td>MYC</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>JAG2</td>\n",
       "      <td>NOTCH1</td>\n",
       "      <td>DLL1</td>\n",
       "      <td>AXIN2</td>\n",
       "      <td>PSEN2</td>\n",
       "      <td>FZD1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   0    \\\n",
       "0     HALLMARK_TNFA_SIGNALING_VIA_NFKB   \n",
       "1                     HALLMARK_HYPOXIA   \n",
       "2     HALLMARK_CHOLESTEROL_HOMEOSTASIS   \n",
       "3             HALLMARK_MITOTIC_SPINDLE   \n",
       "4  HALLMARK_WNT_BETA_CATENIN_SIGNALING   \n",
       "\n",
       "                                                 1        2        3      4    \\\n",
       "0  http://www.broadinstitute.org/gsea/msigdb/card...     JUNB    CXCL2   ATF3   \n",
       "1  http://www.broadinstitute.org/gsea/msigdb/card...     PGK1     PDK1   GBE1   \n",
       "2  http://www.broadinstitute.org/gsea/msigdb/card...     FDPS  CYP51A1   IDI1   \n",
       "3  http://www.broadinstitute.org/gsea/msigdb/card...  ARHGEF2   CLASP1  KIF11   \n",
       "4  http://www.broadinstitute.org/gsea/msigdb/card...      MYC   CTNNB1   JAG2   \n",
       "\n",
       "      5        6      7       8      9     ...      192     193       194  \\\n",
       "0  NFKBIA  TNFAIP3  PTGS2   CXCL1   IER3   ...     EIF1    BMP2     DUSP4   \n",
       "1    PFKL    ALDOA   ENO2    PGM1  NDRG1   ...    HDLBP   ILVBL      NCAN   \n",
       "2   FDFT1    DHCR7   SQLE  HMGCS1  NSDHL   ...      NaN     NaN       NaN   \n",
       "3   KIF23     ALS2   ARF6   MYO9B   MYH9   ...    RHOT2  SORBS2  CDC42EP1   \n",
       "4  NOTCH1     DLL1  AXIN2   PSEN2   FZD1   ...      NaN     NaN       NaN   \n",
       "\n",
       "      195     196     197       198      199       200     201  \n",
       "0  PDLIM5  ICOSLG   GFPT2      KLF2      TNC  SERPINB8    MXD1  \n",
       "1    TGM2    ETS1   HOXB9  SELENBP1    FOSL2   SULT2B1   TGFB3  \n",
       "2     NaN     NaN     NaN       NaN      NaN       NaN     NaN  \n",
       "3     VCL   CLIP1  STK38L     YWHAE  RAPGEF5     CEP72  CSNK1D  \n",
       "4     NaN     NaN     NaN       NaN      NaN       NaN     NaN  \n",
       "\n",
       "[5 rows x 202 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Importing pathway enrichment analysis for 50 Cancer Hallmark pathways http://software.broadinstitute.org/gsea/msigdb/collections.jsp#H\n",
    "#In gmt format each gene set is a row and is described by a name, a description, and the genes in the gene set\n",
    "import pandas as pd\n",
    "\n",
    "H50 = pd.read_csv('h.all.v6.2.symbols.gmt', header = None, sep='\\t')\n",
    "H50.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Add H50 gen sets as a categorical variable "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4387"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Different genes involved in 50 selected signaling pathways \n",
    "genePath_set = pd.unique(H50.loc[:,2:].values.ravel('K'))\n",
    "genePath_set.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(50, 202)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H50.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# log2(TPM+0.001) expression values computed per gene\n",
    "# Uncomment to regenerate\n",
    "#expression = dat['expression'].value\n",
    "#def f(x):\n",
    "    # return log2(TPM+0.001)\n",
    "    #return np.log2(x +0.001 ) \n",
    "#expression = np.vectorize(f)(expression)\n",
    "#expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#from scipy import stats\n",
    "#stats.describe(expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#np.save('log2expression.npy', expression)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[20.47235172, 18.74693086, 18.27246577, ..., 18.2062322 ,\n",
       "        14.86607153, 15.44853581],\n",
       "       [14.81808287, 14.54574888,  7.58497001, ..., 19.94510372,\n",
       "        17.2521942 , 16.72561923],\n",
       "       [12.07748369, 13.24659231, 17.23494838, ..., 20.33197097,\n",
       "        18.41254918, 17.94941639],\n",
       "       ...,\n",
       "       [13.36768789, 14.01959082,  7.78136627, ..., 18.61939969,\n",
       "        18.22982804, 16.39110932],\n",
       "       [14.78289591, 15.66719475,  7.29462994, ..., 20.84872608,\n",
       "        19.33218519, 18.49566769],\n",
       "       [15.30299588, 15.28962272,  5.58499256, ..., 19.59753249,\n",
       "        18.876445  , 18.25677075]])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = np.load('log2expression.npy')\n",
    "l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "      <td>11284.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>14.511335</td>\n",
       "      <td>14.658286</td>\n",
       "      <td>7.818664</td>\n",
       "      <td>21.170343</td>\n",
       "      <td>16.323402</td>\n",
       "      <td>11.515712</td>\n",
       "      <td>10.927732</td>\n",
       "      <td>3.512874</td>\n",
       "      <td>16.375104</td>\n",
       "      <td>4.546174</td>\n",
       "      <td>...</td>\n",
       "      <td>16.317508</td>\n",
       "      <td>16.803166</td>\n",
       "      <td>14.481343</td>\n",
       "      <td>16.262297</td>\n",
       "      <td>17.164771</td>\n",
       "      <td>11.034740</td>\n",
       "      <td>17.568689</td>\n",
       "      <td>19.878162</td>\n",
       "      <td>18.062571</td>\n",
       "      <td>17.549486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2.262990</td>\n",
       "      <td>1.781519</td>\n",
       "      <td>7.636277</td>\n",
       "      <td>1.819773</td>\n",
       "      <td>1.530948</td>\n",
       "      <td>5.025257</td>\n",
       "      <td>2.047451</td>\n",
       "      <td>7.174155</td>\n",
       "      <td>1.840625</td>\n",
       "      <td>7.976631</td>\n",
       "      <td>...</td>\n",
       "      <td>1.238517</td>\n",
       "      <td>1.550544</td>\n",
       "      <td>1.295317</td>\n",
       "      <td>1.108982</td>\n",
       "      <td>0.948553</td>\n",
       "      <td>4.794702</td>\n",
       "      <td>1.120799</td>\n",
       "      <td>1.148660</td>\n",
       "      <td>1.065064</td>\n",
       "      <td>0.987241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>4.906939</td>\n",
       "      <td>4.392386</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>10.540129</td>\n",
       "      <td>8.876520</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>6.988696</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>9.238407</td>\n",
       "      <td>8.906894</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>7.584970</td>\n",
       "      <td>11.969387</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>10.388018</td>\n",
       "      <td>14.118455</td>\n",
       "      <td>12.988152</td>\n",
       "      <td>11.607331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>13.115125</td>\n",
       "      <td>13.612897</td>\n",
       "      <td>6.247946</td>\n",
       "      <td>20.103902</td>\n",
       "      <td>15.396688</td>\n",
       "      <td>8.640249</td>\n",
       "      <td>10.006678</td>\n",
       "      <td>3.322072</td>\n",
       "      <td>15.161555</td>\n",
       "      <td>4.807406</td>\n",
       "      <td>...</td>\n",
       "      <td>15.450841</td>\n",
       "      <td>15.698264</td>\n",
       "      <td>13.734075</td>\n",
       "      <td>15.639362</td>\n",
       "      <td>16.617780</td>\n",
       "      <td>9.012625</td>\n",
       "      <td>16.962933</td>\n",
       "      <td>19.163031</td>\n",
       "      <td>17.430391</td>\n",
       "      <td>16.973854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>14.377075</td>\n",
       "      <td>14.614969</td>\n",
       "      <td>7.936644</td>\n",
       "      <td>21.240994</td>\n",
       "      <td>16.313299</td>\n",
       "      <td>10.918491</td>\n",
       "      <td>11.040290</td>\n",
       "      <td>6.584978</td>\n",
       "      <td>16.524121</td>\n",
       "      <td>7.584970</td>\n",
       "      <td>...</td>\n",
       "      <td>16.299789</td>\n",
       "      <td>16.862601</td>\n",
       "      <td>14.602264</td>\n",
       "      <td>16.340676</td>\n",
       "      <td>17.227212</td>\n",
       "      <td>11.866893</td>\n",
       "      <td>17.655066</td>\n",
       "      <td>19.896538</td>\n",
       "      <td>18.149090</td>\n",
       "      <td>17.624513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>15.611969</td>\n",
       "      <td>15.556147</td>\n",
       "      <td>11.956830</td>\n",
       "      <td>22.350001</td>\n",
       "      <td>17.271283</td>\n",
       "      <td>14.611140</td>\n",
       "      <td>12.092427</td>\n",
       "      <td>8.029291</td>\n",
       "      <td>17.722402</td>\n",
       "      <td>9.205794</td>\n",
       "      <td>...</td>\n",
       "      <td>17.199718</td>\n",
       "      <td>17.988194</td>\n",
       "      <td>15.369069</td>\n",
       "      <td>16.953637</td>\n",
       "      <td>17.761895</td>\n",
       "      <td>14.115450</td>\n",
       "      <td>18.333242</td>\n",
       "      <td>20.566523</td>\n",
       "      <td>18.753259</td>\n",
       "      <td>18.193407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>25.744343</td>\n",
       "      <td>23.457815</td>\n",
       "      <td>21.781484</td>\n",
       "      <td>27.526837</td>\n",
       "      <td>22.749258</td>\n",
       "      <td>25.852982</td>\n",
       "      <td>18.892860</td>\n",
       "      <td>14.192447</td>\n",
       "      <td>21.365289</td>\n",
       "      <td>20.369699</td>\n",
       "      <td>...</td>\n",
       "      <td>20.524677</td>\n",
       "      <td>21.309852</td>\n",
       "      <td>17.978504</td>\n",
       "      <td>20.607729</td>\n",
       "      <td>20.899094</td>\n",
       "      <td>21.113483</td>\n",
       "      <td>22.043753</td>\n",
       "      <td>24.416677</td>\n",
       "      <td>22.150687</td>\n",
       "      <td>21.651597</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25150 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A1BG      A1BG-AS1          A1CF           A2M       A2M-AS1  \\\n",
       "count  11284.000000  11284.000000  11284.000000  11284.000000  11284.000000   \n",
       "mean      14.511335     14.658286      7.818664     21.170343     16.323402   \n",
       "std        2.262990      1.781519      7.636277      1.819773      1.530948   \n",
       "min        4.906939      4.392386     -9.965784     10.540129      8.876520   \n",
       "25%       13.115125     13.612897      6.247946     20.103902     15.396688   \n",
       "50%       14.377075     14.614969      7.936644     21.240994     16.313299   \n",
       "75%       15.611969     15.556147     11.956830     22.350001     17.271283   \n",
       "max       25.744343     23.457815     21.781484     27.526837     22.749258   \n",
       "\n",
       "              A2ML1         A2MP1       A3GALT2        A4GALT         A4GNT  \\\n",
       "count  11284.000000  11284.000000  11284.000000  11284.000000  11284.000000   \n",
       "mean      11.515712     10.927732      3.512874     16.375104      4.546174   \n",
       "std        5.025257      2.047451      7.174155      1.840625      7.976631   \n",
       "min       -9.965784     -9.965784     -9.965784      6.988696     -9.965784   \n",
       "25%        8.640249     10.006678      3.322072     15.161555      4.807406   \n",
       "50%       10.918491     11.040290      6.584978     16.524121      7.584970   \n",
       "75%       14.611140     12.092427      8.029291     17.722402      9.205794   \n",
       "max       25.852982     18.892860     14.192447     21.365289     20.369699   \n",
       "\n",
       "           ...             ZWILCH         ZWINT          ZXDA          ZXDB  \\\n",
       "count      ...       11284.000000  11284.000000  11284.000000  11284.000000   \n",
       "mean       ...          16.317508     16.803166     14.481343     16.262297   \n",
       "std        ...           1.238517      1.550544      1.295317      1.108982   \n",
       "min        ...           9.238407      8.906894     -9.965784      7.584970   \n",
       "25%        ...          15.450841     15.698264     13.734075     15.639362   \n",
       "50%        ...          16.299789     16.862601     14.602264     16.340676   \n",
       "75%        ...          17.199718     17.988194     15.369069     16.953637   \n",
       "max        ...          20.524677     21.309852     17.978504     20.607729   \n",
       "\n",
       "               ZXDC        ZYG11A        ZYG11B           ZYX         ZZEF1  \\\n",
       "count  11284.000000  11284.000000  11284.000000  11284.000000  11284.000000   \n",
       "mean      17.164771     11.034740     17.568689     19.878162     18.062571   \n",
       "std        0.948553      4.794702      1.120799      1.148660      1.065064   \n",
       "min       11.969387     -9.965784     10.388018     14.118455     12.988152   \n",
       "25%       16.617780      9.012625     16.962933     19.163031     17.430391   \n",
       "50%       17.227212     11.866893     17.655066     19.896538     18.149090   \n",
       "75%       17.761895     14.115450     18.333242     20.566523     18.753259   \n",
       "max       20.899094     21.113483     22.043753     24.416677     22.150687   \n",
       "\n",
       "               ZZZ3  \n",
       "count  11284.000000  \n",
       "mean      17.549486  \n",
       "std        0.987241  \n",
       "min       11.607331  \n",
       "25%       16.973854  \n",
       "50%       17.624513  \n",
       "75%       18.193407  \n",
       "max       21.651597  \n",
       "\n",
       "[8 rows x 25150 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#associate gene names with RNA expression\n",
    "def f(x):\n",
    "    return x.decode(\"utf-8\")\n",
    "genes = np.vectorize(f)(meta['genes'].value)\n",
    "df = pd.DataFrame(l, columns = genes)\n",
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genePath_set' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-7426098d74d1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#One hot 50 signaling pathways for each instance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#1. Select only genes involved in pathways\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenePath_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mgenePath_I_geneRNA\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintersection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenePath_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenePath_I_geneRNA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genePath_set' is not defined"
     ]
    }
   ],
   "source": [
    "#One hot 50 signaling pathways for each instance\n",
    "#1. Select only genes involved in pathways\n",
    "print(genePath_set)\n",
    "genePath_I_geneRNA= set(df.columns).intersection(set(genePath_set))\n",
    "print(len(genePath_I_geneRNA))\n",
    "df_s = df[list(genePath_I_geneRNA)]\n",
    "print(\"RNA exp gene matrix reduced to only genes in pathways\")\n",
    "print(df_s.shape)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_s' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-f17ba5037795>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrna\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mdf_s\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'H50_onehot'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_s\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;31m#Merge H50 onehot with the full RNA expression matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'H50_onehot'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_s\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'H50_onehot'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_s' is not defined"
     ]
    }
   ],
   "source": [
    "#2. Add result of pathway enrichment analysis for 50 Cancer Hallmark pathways as categorical variables: \n",
    "#Array of 50 boolean\n",
    "#If all genes involved on each signaling pathway from H50 are RNA log2 expressed > 0 then this pathway is assigned True in the sample\n",
    "def f(row):\n",
    "    a = []\n",
    "    for index, p in H50.iterrows():\n",
    "        #print(p)\n",
    "        #exclude from pathway genes 'p' those that are not in genePath_I_geneRNA (Total 83)\n",
    "        p_s = set(p[2:]).intersection(genePath_I_geneRNA)\n",
    "        rna = row[list(p_s)]\n",
    "        #print (rna)\n",
    "        a.append(np.all(rna.values > 0))\n",
    "    return np.array(a)\n",
    "df_s['H50_onehot'] = df_s.apply(f, axis=1)\n",
    "#Merge H50 onehot with the full RNA expression matrix\n",
    "df['H50_onehot'] = df_s['H50_onehot'] \n",
    "df.to_pickle(\"RNAexp_H50onehot.pkl\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>H50_onehot</th>\n",
       "      <th>cancertype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20.472352</td>\n",
       "      <td>18.746931</td>\n",
       "      <td>18.272466</td>\n",
       "      <td>19.290201</td>\n",
       "      <td>15.257498</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>9.299210</td>\n",
       "      <td>5.643885</td>\n",
       "      <td>13.595724</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>11.805341</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>15.517669</td>\n",
       "      <td>9.278452</td>\n",
       "      <td>15.699383</td>\n",
       "      <td>18.206232</td>\n",
       "      <td>14.866072</td>\n",
       "      <td>15.448536</td>\n",
       "      <td>[False, False, False, True, False, False, Fals...</td>\n",
       "      <td>b'Liver Hepatocellular Carcinoma'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14.818083</td>\n",
       "      <td>14.545749</td>\n",
       "      <td>7.584970</td>\n",
       "      <td>21.170981</td>\n",
       "      <td>17.082887</td>\n",
       "      <td>13.392988</td>\n",
       "      <td>10.586841</td>\n",
       "      <td>3.322072</td>\n",
       "      <td>17.862692</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>13.460072</td>\n",
       "      <td>15.752668</td>\n",
       "      <td>16.611414</td>\n",
       "      <td>12.744834</td>\n",
       "      <td>16.470659</td>\n",
       "      <td>19.945104</td>\n",
       "      <td>17.252194</td>\n",
       "      <td>16.725619</td>\n",
       "      <td>[True, False, True, True, True, True, False, T...</td>\n",
       "      <td>b'Prostate Adenocarcinoma'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12.077484</td>\n",
       "      <td>13.246592</td>\n",
       "      <td>17.234948</td>\n",
       "      <td>20.949892</td>\n",
       "      <td>14.822969</td>\n",
       "      <td>6.700454</td>\n",
       "      <td>10.472692</td>\n",
       "      <td>7.321937</td>\n",
       "      <td>16.977783</td>\n",
       "      <td>7.044405</td>\n",
       "      <td>...</td>\n",
       "      <td>15.637785</td>\n",
       "      <td>17.493347</td>\n",
       "      <td>16.861135</td>\n",
       "      <td>8.839207</td>\n",
       "      <td>17.636320</td>\n",
       "      <td>20.331971</td>\n",
       "      <td>18.412549</td>\n",
       "      <td>17.949416</td>\n",
       "      <td>[True, False, True, True, True, True, False, T...</td>\n",
       "      <td>b'Rectum Adenocarcinoma'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24.457885</td>\n",
       "      <td>22.684448</td>\n",
       "      <td>19.907039</td>\n",
       "      <td>23.860669</td>\n",
       "      <td>19.778088</td>\n",
       "      <td>7.918869</td>\n",
       "      <td>13.952559</td>\n",
       "      <td>1.000721</td>\n",
       "      <td>16.081317</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>12.749870</td>\n",
       "      <td>15.258603</td>\n",
       "      <td>16.945592</td>\n",
       "      <td>12.561527</td>\n",
       "      <td>18.267911</td>\n",
       "      <td>16.341345</td>\n",
       "      <td>17.340346</td>\n",
       "      <td>17.422426</td>\n",
       "      <td>[False, False, True, True, True, True, False, ...</td>\n",
       "      <td>b'Liver Hepatocellular Carcinoma'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11.676398</td>\n",
       "      <td>12.256504</td>\n",
       "      <td>8.076821</td>\n",
       "      <td>19.457004</td>\n",
       "      <td>14.529736</td>\n",
       "      <td>17.318905</td>\n",
       "      <td>8.209458</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>17.581149</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>13.757911</td>\n",
       "      <td>15.446534</td>\n",
       "      <td>17.096242</td>\n",
       "      <td>12.186734</td>\n",
       "      <td>17.651276</td>\n",
       "      <td>20.299427</td>\n",
       "      <td>17.006589</td>\n",
       "      <td>17.689343</td>\n",
       "      <td>[False, False, True, True, False, True, False,...</td>\n",
       "      <td>b'Bladder Urothelial Carcinoma'</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25152 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        A1BG   A1BG-AS1       A1CF        A2M    A2M-AS1      A2ML1  \\\n",
       "0  20.472352  18.746931  18.272466  19.290201  15.257498  -9.965784   \n",
       "1  14.818083  14.545749   7.584970  21.170981  17.082887  13.392988   \n",
       "2  12.077484  13.246592  17.234948  20.949892  14.822969   6.700454   \n",
       "3  24.457885  22.684448  19.907039  23.860669  19.778088   7.918869   \n",
       "4  11.676398  12.256504   8.076821  19.457004  14.529736  17.318905   \n",
       "\n",
       "       A2MP1   A3GALT2     A4GALT     A4GNT  \\\n",
       "0   9.299210  5.643885  13.595724 -9.965784   \n",
       "1  10.586841  3.322072  17.862692 -9.965784   \n",
       "2  10.472692  7.321937  16.977783  7.044405   \n",
       "3  13.952559  1.000721  16.081317 -9.965784   \n",
       "4   8.209458 -9.965784  17.581149 -9.965784   \n",
       "\n",
       "                 ...                       ZXDA       ZXDB       ZXDC  \\\n",
       "0                ...                  11.805341  13.000000  15.517669   \n",
       "1                ...                  13.460072  15.752668  16.611414   \n",
       "2                ...                  15.637785  17.493347  16.861135   \n",
       "3                ...                  12.749870  15.258603  16.945592   \n",
       "4                ...                  13.757911  15.446534  17.096242   \n",
       "\n",
       "      ZYG11A     ZYG11B        ZYX      ZZEF1       ZZZ3  \\\n",
       "0   9.278452  15.699383  18.206232  14.866072  15.448536   \n",
       "1  12.744834  16.470659  19.945104  17.252194  16.725619   \n",
       "2   8.839207  17.636320  20.331971  18.412549  17.949416   \n",
       "3  12.561527  18.267911  16.341345  17.340346  17.422426   \n",
       "4  12.186734  17.651276  20.299427  17.006589  17.689343   \n",
       "\n",
       "                                          H50_onehot  \\\n",
       "0  [False, False, False, True, False, False, Fals...   \n",
       "1  [True, False, True, True, True, True, False, T...   \n",
       "2  [True, False, True, True, True, True, False, T...   \n",
       "3  [False, False, True, True, True, True, False, ...   \n",
       "4  [False, False, True, True, False, True, False,...   \n",
       "\n",
       "                          cancertype  \n",
       "0  b'Liver Hepatocellular Carcinoma'  \n",
       "1         b'Prostate Adenocarcinoma'  \n",
       "2           b'Rectum Adenocarcinoma'  \n",
       "3  b'Liver Hepatocellular Carcinoma'  \n",
       "4    b'Bladder Urothelial Carcinoma'  \n",
       "\n",
       "[5 rows x 25152 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Add the ground truth: tumor type for each instance\n",
    "#Load processed data\n",
    "df = pd.read_pickle('RNAexp_H50onehot.pkl')\n",
    "df.loc[:,'cancertype']= pd.Series(meta['cancertype'].value)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11284, 25202)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(11284, 25201)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H50_categories = pd.DataFrame(df.H50_onehot.tolist(), columns = H50.loc[:,0].values)\n",
    "df_categories = df.join(H50_categories)\n",
    "df_categories.H50_onehot\n",
    "print(df_categories.shape)\n",
    "df_categories.to_pickle(\"RNAexp_H50cat_tumorTypes.pkl\")\n",
    "df_categories = df_categories.drop(columns=['H50_onehot'])\n",
    "df_categories.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    11284.000000\n",
       "mean        15.014002\n",
       "std          9.301482\n",
       "min          0.000000\n",
       "25%          7.000000\n",
       "50%         14.000000\n",
       "75%         23.000000\n",
       "max         32.000000\n",
       "Name: cancertype, dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Use FastAI \n",
    "from fastai import *\n",
    "from fastai.tabular import * \n",
    "\n",
    "procs = [FillMissing, Categorify, Normalize]\n",
    "dep_var = 'cancertype'\n",
    "cat_names = H50.loc[:,0].values #all other fields not in dep_var or cat_names are handled as continous variables\n",
    "valid_idx = range(len(df_categories)-3000, len(df_categories))\n",
    "df_categories['cancertype'] = pd.Categorical(df_categories.cancertype)\n",
    "df_categories['cancertype'] = df_categories.cancertype.cat.codes\n",
    "df_categories['cancertype'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time: 02:40\n",
      "epoch  train_loss  valid_loss  accuracy\n",
      "1      0.503842    0.241396    0.923333  (00:04)\n",
      "2      0.227772    0.213783    0.927000  (00:04)\n",
      "3      0.127209    0.185534    0.940333  (00:04)\n",
      "4      0.090626    0.199434    0.939333  (00:04)\n",
      "5      0.077104    0.212278    0.934000  (00:04)\n",
      "6      0.049125    0.217233    0.950000  (00:04)\n",
      "7      0.037340    0.340264    0.932000  (00:04)\n",
      "8      0.046956    0.250232    0.947000  (00:04)\n",
      "9      0.032279    0.284934    0.935000  (00:04)\n",
      "10     0.028284    0.307077    0.931000  (00:04)\n",
      "11     0.026738    0.246928    0.948000  (00:04)\n",
      "12     0.025021    0.407419    0.933333  (00:04)\n",
      "13     0.022676    0.260365    0.951000  (00:05)\n",
      "14     0.018009    0.257445    0.951333  (00:04)\n",
      "15     0.025444    0.249987    0.949000  (00:04)\n",
      "16     0.013448    0.248825    0.956000  (00:04)\n",
      "17     0.019259    0.323733    0.941333  (00:04)\n",
      "18     0.009642    0.237344    0.950000  (00:04)\n",
      "19     0.023478    0.252720    0.950000  (00:04)\n",
      "20     0.019434    0.250747    0.954667  (00:04)\n",
      "21     0.009916    0.244671    0.952000  (00:04)\n",
      "22     0.011961    0.299021    0.945000  (00:04)\n",
      "23     0.012448    0.320385    0.944000  (00:05)\n",
      "24     0.009961    0.345773    0.951000  (00:04)\n",
      "25     0.010449    0.254550    0.956667  (00:04)\n",
      "26     0.012916    0.307155    0.950333  (00:04)\n",
      "27     0.011771    0.265072    0.950667  (00:04)\n",
      "28     0.008584    0.289053    0.948333  (00:04)\n",
      "29     0.013058    0.267880    0.951667  (00:04)\n",
      "30     0.007896    0.370104    0.953000  (00:04)\n",
      "31     0.010998    0.393231    0.938000  (00:04)\n",
      "32     0.007198    0.274654    0.949000  (00:04)\n",
      "33     0.011277    0.275129    0.949667  (00:04)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "data = TabularDataBunch.from_df('./', df_categories, dep_var=dep_var, valid_idx=valid_idx, procs=procs, cat_names=cat_names)\n",
    "\n",
    "#print(data.train_ds.cont_names)  # `cont_names` defaults to: set(df)-set(cat_names)-{dep_var}\n",
    "\n",
    "learn = tabular_learner(data, layers=[500,50], metrics=accuracy, ps=[.2,.1])\n",
    "learn.fit(33, 1e-2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python (gen)",
   "language": "python",
   "name": "gen"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
